Song Selector in Native Code

epeach opened the pull request:
Ready for Review

Updates new runtime with Song Selector, song can be chosen, played, and the time retrieved. Does not handle cues.

Song meta data is copied twice in this PR. This copied data will get removed once we retrieve the meta data from S3.

cpirich commented on apps/src/Sound.js:
```diff
@@ -308,6 +316,7 @@ Sound.prototype.onSoundLoaded = function () {
   if (this.onLoad) {
     this.onLoad();
   }
+  this.isLoaded_ = true;
```
I would move this line above the `onLoad()` callback. It seems better that if the callback code called the new `isLoaded()`, it would return `true` at this point.

cpirich commented on apps/src/Sounds.js:
```diff
@@ -259,6 +259,16 @@ Sounds.prototype.play = function (soundId, options) {
   if (sound) {
     sound.play(options);
   }
+  this.audioStartedTime = this.audioContext.currentTime;
```
I don't think this will work for HTML5 audio, where we don't have an `audioContext`, but we do have an `audioElement`.

cpirich commented on apps/src/Sounds.js:
```diff
@@ -259,6 +259,16 @@ Sounds.prototype.play = function (soundId, options) {
+    return this.audioContext.currentTime - this.audioStartedTime;
+  } else {
+    return 0;
+  }
```
For the HTML5 audio case, you'll probably be able to use `audioElement.currentTime`, but I'm not sure how accurate that will be. We may be better off capturing `this.audioStartedTime` for both type of audio playback using something like this: https://developer.mozilla.org/en-US/docs/Web/API/Performance/now

cpirich replied:
Also, this seems odd to place on the `Sounds` object, as we allow multiple sounds to play simultaneously.

cpirich commented on apps/src/dance/Dance.js:
```diff
@@ -196,6 +230,11 @@ Dance.prototype.onReportComplete = function (response) {
+  if (!Sounds.getSingleton().isLoaded(getStore().getState().selectedSong)) {
+    return;
+  }
+
```
This seems odd. I thought we could just start when the user pressed Run and then wait until the song was fully loaded before animating. That's how the prototype I created worked with the playSound() callback.

joshlory replied:
This was my bad, I suggested following the pattern we use in Craft (that just bails if the runtime isn't loaded yet).

https://github.com/code-dot-org/code-dot-org/blob/67eb8face9fcea5e287edd91c21a85170b38f761/apps/src/craft/simple/craft.js#L561-L564

If we delay the start, would we need to add a spinner and protect against multiple clicks?  Prefer to address this in a follow-up PR.

joshlory replied:
Ah, I didn't read your top-level review comment first.  What you're saying makes sense if we start p5 right away, but only draw a loader until the callback is fired!

We should go with whatever is simpler to get this set of changes merged for now...

cpirich replied:
Since this song could take a while to load over slow connections, it seems like we should not have "dead" Run button presses where nothing happens. I'm happy to help as needed to get this in - just let me know...

cpirich replied:
Also, I don't think a loader is really required - the Run button will at least toggle to Reset and the screen can still still have a background effect active since running will have started. It's just that the dancer moves shouldn't start until the song is playing...

epeach commented:
Thanks for the feedback, @cpirich and @joshlory ! PTAL when you get a chance. I updated to use the playSound and removed the 'early out' for when the song isn't yet loaded (and removed the associated helpers that I had added). This PR doesn't handle the case when someone presses 'Play' before the song is loaded. In the interest of getting is PR merged ASAP, I believe we can add that in a follow-up PR.

epeach commented:
Also, a known non-critical issue - when the first song is loaded, it displays a large number of measures. Resolves once a program is run. Considering non-ship-blocking.

cpirich commented:
Thanks @epeach - you definitely addressed my primary feedback. It would be great to see us handle the case where the song isn't ready. And also, I think you'll need to reset the `songStartTime` during a Reset/Run. But, generally - I think this is much better. The `register()` calls are a hint to the system to start loading the song, but aren't strictly required. We probably need to do a little more work to ensure that a `playSound()` call that is made after a register/preload call that hasn't yet completed will still queue up a "play" when the load completes. I can take a look at that this week to make sure this is air tight.
